internal/iapl/policy.go

```diff
@@ -257,20 +399,205 @@ func (v *policy) expandActionBindings() {
+		role.Relationships = permsRel
+	} else {
+		role = ResourceType{
+			Name:          v.p.RBAC.RoleResource,
+			Relationships: permsRel,
+		}
+	}
+
+	v.rt[role.Name] = role
+
+	// 5. create a list of role owners
+	for _, owner := range v.p.RBAC.RoleOwners {
+		v.roleOwners[owner] = struct{}{}
+	}
+}
+
+func (v *policy) expandRoleBinding() {
+	// 1. create relationship to role
+	role := Relationship{
+		Relation: RolebindingRoleRelation,
+		TargetTypes: []types.TargetType{
+			{Name: v.p.RBAC.RoleResource},
+		},
+	}
+
+	// 2. create relationship to subjects
+	subjects := Relationship{
+		Relation:    RolebindingSubjectRelation,
+		TargetTypes: v.p.RBAC.RoleBindingSubjects,
+	}
+
+	// 3. create a list of action-bindings representing permissions for all the
+	// actions
+	actionbindings := make([]ActionBinding, len(v.ac))
+	i := 0
+
+	for actionName := range v.ac {
+		ab := ActionBinding{
+			ActionName: actionName,
+			TypeName:   v.p.RBAC.RoleBindingResource,
+			ConditionSets: []types.ConditionSet{
+				{
+					Conditions: []types.Condition{
+						{
+							RelationshipAction: &types.ConditionRelationshipAction{
+								Relation:   RolebindingRoleRelation,
+								ActionName: actionName + "_rel",
+							},
+						},
+					},
+				},
+				{
+					Conditions: []types.Condition{
+						{RelationshipAction: &types.ConditionRelationshipAction{Relation: RolebindingSubjectRelation}},
+					},
+				},
+			},
+		}
+
+		actionbindings[i] = ab
+		i++
+	}
+
+	v.bn = append(v.bn, actionbindings...)
+
+	// 4. create role-binding resource type
+	var rolebinding ResourceType
+
+	if _, ok := v.rt[v.p.RBAC.RoleBindingResource]; ok {
+		rolebinding = v.rt[v.p.RBAC.RoleBindingResource]
+		rolebinding.Relationships = []Relationship{role, subjects}
+	} else {
+		rolebinding = ResourceType{
+			Name:          v.p.RBAC.RoleBindingResource,
+			Relationships: []Relationship{role, subjects},
+		}
+	}
+
+	v.rt[v.p.RBAC.RoleBindingResource] = rolebinding
+}
+
 func (v *policy) expandResourceTypes() {
 	for name, resourceType := range v.rt {
 		for i, rel := range resourceType.Relationships {
 			var typeNames []string
 
+			targettypes := rel.TargetTypes
+
 			for _, typeName := range rel.TargetTypeNames {
 				if u, ok := v.un[typeName]; ok {
-					typeNames = append(typeNames, u.ResourceTypeNames...)
+					if len(u.ResourceTypes) > 0 {
+						targettypes = append(targettypes, u.ResourceTypes...)
+					} else {
+						typeNames = append(typeNames, u.ResourceTypeNames...)
+					}
 				} else {
 					typeNames = append(typeNames, typeName)
 				}
 			}
 
+			for _, tn := range typeNames {
+				targettypes = append(targettypes, types.TargetType{Name: tn})
+			}
+
 			resourceType.Relationships[i].TargetTypeNames = typeNames
+			resourceType.Relationships[i].TargetTypes = targettypes
+		}
+
+		// role-binding V2 configs
+		availRoleConditions := []Condition{}
+
+		// if resource type is a role-owner, add the role-relationship to the
+		// resource
```

Comment:
Is this the right place to do this? It might be, but in general I think we'll have a less difficult time if permissions-api defines Role and RoleBinding resource types as early as possible and adds them to the policy.

In any case, doing that in a separate method would be better anyway.